### PR TITLE
FINERACT-1926: Fineract Asset Externalization - GET Pagination

### PR DESCRIPTION
**Pull request by b0c1**
- [x] Added Pagination to response
- [x] Added limit and offset to the request

## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/service/Page.java [hunk 1]
```diff
@@ -38,4 +38,8 @@ public int getTotalFilteredRecords() {
     public List<E> getPageItems() {
         return this.pageItems;
     }
+
+    public static <E> Page<E> convertFromSpringPage(final org.springframework.data.domain.Page<E> springPage) {
```
Comment (galovics) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/service/Page.java]: Let's remove this.

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/ExternalAssetOwnersReadServiceImpl.java [hunk 2]
```diff
@@ -37,15 +37,27 @@ public class ExternalAssetOwnersReadServiceImpl implements ExternalAssetOwnersRe
     private final ExternalAssetOwnersTransferMapper mapper;
 
     @Override
-    public List<ExternalTransferData> retrieveTransferData(Long loanId, String externalLoanId, String transferExternalId) {
-        List<ExternalAssetOwnerTransfer> result = new ArrayList<>();
+    public Page<ExternalTransferData> retrieveTransferData(Long loanId, String externalLoanId, String transferExternalId, Integer offset,
+            Integer limit) {
+        org.springframework.data.domain.Page<ExternalAssetOwnerTransfer> result;
```
Comment (galovics) [inline fineract-investor/src/main/java/org/apache/fineract/investor/service/ExternalAssetOwnersReadServiceImpl.java]: No need for qualified name.

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/ExternalAssetOwnersReadServiceImpl.java [hunk 3]
```diff
@@ -37,15 +37,27 @@ public class ExternalAssetOwnersReadServiceImpl implements ExternalAssetOwnersRe
...
+        if (offset == null) {
+            offset = 0;
+        }
+        if (limit == null) {
+            limit = 100;
+        }
+        PageRequest pageRequest = PageRequest.of(offset, limit);
         if (loanId != null) {
-            result.addAll(externalAssetOwnerTransferRepository.findAllByLoanId(loanId));
+            result = externalAssetOwnerTransferRepository.findAllByLoanId(loanId, pageRequest);
```
Comment (galovics) [inline fineract-investor/src/main/java/org/apache/fineract/investor/service/ExternalAssetOwnersReadServiceImpl.java:52]: I think for all these repository calls where we use pageable we are missing a crucial thing, a consistent ordering of the result. Let's do an ID based ordering at least for the query to produce consistent results between runs.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/InitiateExternalAssetOwnerTransferTest.java [hunk 4]
```diff
@@ -114,17 +115,16 @@ public void saleActiveLoanToExternalAssetOwner() {
             assertEquals(responseMap.get("resourceExternalId"), transferExternalId);
 
             String retrieveResponse = externalAssetOwnerHelper.retrieveTransferByLoanId(loanID.longValue());
-            Type retrieveType = new TypeToken<List<Map<String, Object>>>() {}.getType();
-            List<Map<String, Object>> retrieveResponseMap = new Gson().fromJson(retrieveResponse, retrieveType);
-            assertEquals(1, retrieveResponseMap.size());
-            assertEquals(retrieveResponseMap.get(0).get("transferExternalId"), transferExternalId);
-            assertEquals(retrieveResponseMap.get(0).get("status"), "PENDING");
+            Type retrieveType = new TypeToken<Page<Map<String, Object>>>() {}.getType();
+            Page<Map<String, Object>> retrieveResponseMap = new Gson().fromJson(retrieveResponse, retrieveType);
```
Comment (galovics) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/InitiateExternalAssetOwnerTransferTest.java]: I think we need an additional test case to verify that paging is working. Don't you think?

**Review by galovics — CHANGES_REQUESTED** [review]

**Comment by adamsaghy** [comment]
@b0c1 Kindly asking you to check the failing tests!

**Review by galovics — APPROVED** [review]

**Pull request merged**